rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+updates also ensure that users can take advantage of improvements in stability and
+performance.  Outdated teleport installations impose additional burdens on us and on
+our users. Teleport does not currently assist with upgrades in any way, and the burden
+of manual upgrades can be prohibitive.
+
+Reducing the friction of upgrades is beneficial both in terms of security and user
+experience. Doing this may also indirectly lower our own support load.
+
+
+## Intro
+
+### Suggested Reading
+
+While not required, it is helpful to have some familiarity with [The Update Framework](https://theupdateframework.com/)
+when reading this RFD. TUF is a flexible framework for securing upgrade systems. It provides a robust framework
+for key rotation, censorship detection, package validation, and much more.
+
+
+### High-Level Goals
+
+1. Maintain or improve the security of teleport installations by keeping them better
+updated and potentially providing more secure paths to upgrade.
+
+2. Improve the experience of teleport upgrade/administration by reducing the need for
+manual intervention.
+
+3. Improve the auditability of teleport clusters by providing insight into, and policy
+enforcement for, the versioning of teleport installations.
+
+4. Support a wide range of use-cases by providing a flexible and extensible set of tools
+with support for things like user-provided upgrade scripts and custom target selection.
+
+5. Offer a simple "batteries included" automatic upgrade option that requires minimal
+configuration and "just works".
+
+
+### Abstract Model Overview
+
+This document proposes an modular system capable of supporting a wide range
+of upgrade strategies, with the intention being that the default or "batteries
+included" upgrade strategy will be implemented primarily as a set of interchangeable
+components which can be swapped out and/or extended.
+
+The proposed system consists of at least the following components:
+
+- `version-directive`: A static resource that describes the desired state of
+versioning across the cluster. The directive includes matchers which allow
+the auth server to match individual teleport instances with both the appropriate
+installation target, and the appropriate installation method. This resource may
+be periodically generated by teleport or by some custom external program. It may
+also be manually created by an administrator.
+
+- `version-controller`: A pluggable component responsible for generating the
+`version-directive` based on some dynamic state (e.g. a server which publishes
+package versions and hashes). A builtin `version-controller` would simply be a
+configuration resource from the user's perspective. A custom/external version
+controller would be any program with the permissions necessary to update the
+`version-directive` resource.
+
+- *version reconciliation loop*: A control loop that runs in the auth server which
```

Comment:
Would it be more accurate to say that the version reconciliation loop runs in the version controller (as opposed to the auth server)?

Reply:
No, the two things are different.  A version controller generates a detailed description of what the desired state is, and what installers to use to get to that state.  The controller could be anything... might run on the auth server, might be plugin, or might even be a human manually writing the directive.
The reconciliation loop is the builtin teleport functionality that orchestrates the details of how the directive gets applied.  E.g. if a version controller generates a directive to the affect of "all servers with label `env:prod` should be running `v1.2.3`, which should be installed using `local-script/my-installer`", the reconciliation loop is what looks at that directive and decides which individual servers actually need to have the installer run, then invokes the installer for those servers.